Support synthetic source for counted_keyword field type

jordan-powers opened the pull request:
This patch updates the `counted_keyword` field type to natively support synthetic source, retrieving its source values from the doc values. We still fallback to the ignore source infrastructure if the `synthetic_source_keep` mode is not `none`.

As a part of this patch, some updates had to be made to the `parseCreateField` logic: 1) to avoid creating the "_count" lucene field if there are no values to store and 2) to handle nested arrays without throwing an exception. These issues were revealed by the synthetic source mapper tests.

Relates to #109796

jordan-powers committed:
Natively support synthetic source for the counted_keyword field type

jordan-powers committed:
Add tests for counted_keyword fields native synthetic source support

jordan-powers committed:
Update CountedKeywordFieldMapper::parseArray to handle nested arrays

jordan-powers committed:
CountedKeywordFieldMapper: fall back to stored source if SourceKeepMode is not NONE

elasticsearchmachine commented:
Pinging @elastic/es-storage-engine (Team:StorageEngine)

jordan-powers committed:
Merge branch 'main' into counted_keyword_synthetic_source

martijnvg commented on x-pack/plugin/mapper-counted-keyword/src/main/java/org/elasticsearch/xpack/countedkeyword/CountedKeywordFieldMapper.java:
```diff
@@ -399,4 +485,16 @@ public FieldMapper.Builder getMergeBuilder() {
+    @Override
+    protected SyntheticSourceSupport syntheticSourceSupport() {
+        var keepMode = sourceKeepMode();
+        if (keepMode.isPresent() && (keepMode.get() == SourceKeepMode.ARRAYS || keepMode.get() == SourceKeepMode.ALL)) {
```
```suggestion
        if (keepMode.isPresent() && (keepMode.get() != SourceKeepMode.NONE) {
```

martijnvg commented on x-pack/plugin/mapper-counted-keyword/src/main/java/org/elasticsearch/xpack/countedkeyword/CountedKeywordFieldMapper.java:
```diff
@@ -348,20 +427,27 @@ protected void parseCreateField(DocumentParserContext context) throws IOExceptio
-        BytesStreamOutput streamOutput = new BytesStreamOutput();
-        streamOutput.writeVIntArray(counts);
-        context.doc().add(new BinaryDocValuesField(countFieldMapper.fullPath(), streamOutput.bytes().toBytesRef()));
+        if (counts.length > 0) {
```
+1 this should work. If count array is empty than there should also be no values. And assumption in `CountedKeywordSortedBinaryDocValues` that if there is a value then there should be a entry in count field should not break.

Maybe instead we can even stop earlier by adding this statement before `I` variable is initialized:

```java
if (values.isEmpty()) {
   return;
}
```

martijnvg reviewed (commented):
Nice work! I left two comments. Maybe also add a yaml test to `./x-pack/plugin/src/yamlRestTest/resources/rest-api-spec.test/counted_keyword` directory?

jordan-powers committed:
Review suggestions

jordan-powers commented:
Thanks for the review @martijnvg! I've made the suggested updates, and I'll start working on a yaml test

martijnvg commented:
>  and I'll start working on a yaml test

I think a test that creates an index with counted keyword (and source mode set to `synthetic`), indexes a document and perform a get or search in order to see that the document's _source with field mapped as counted keyword is as expected.

jordan-powers commented:
Yes, I've mostly got that written, I'm just having trouble figuring out how to run the yaml tests. Keep getting an error:
```
java.lang.RuntimeException: Elasticsearch process died while waiting for ports file. See console output for details.
```

To run, I'm using the command:
```bash
./gradlew :x-pack:plugin:yamlRestTest
```

martijnvg commented:
The build dir in x-pack/plugin should include an es log file. Maybe run with: `./gradlew :x-pack:plugin:yamlRestTest --info`? This prints trailing part of es log file to the console. Note that you should also be to just run your tests with test name filtering, by adding something like: `-Dtests.method="*your_test_name*"`

jordan-powers commented:
Ok, seems like the issue is that it's trying to load a ML plugin, which is compiled in C++ and doesn't support my platform.
```
    [2025-01-15T06:01:54,047][FATAL][o.e.x.c.XPackSettings    ] [yamlRestTest-0] xpack.ml.enabled cannot be set to [true] on [darwin-x86_64]
    [2025-01-15T06:01:54,048][ERROR][o.e.b.Elasticsearch      ] [yamlRestTest-0] fatal exception while booting Elasticsearch java.lang.IllegalStateException: failed to load plugin class [org.elasticsearch.xpack.ml.MachineLearning]
        at org.elasticsearch.server@9.0.0-SNAPSHOT/org.elasticsearch.plugins.PluginsService.loadPlugin(PluginsService.java:521)
        at org.elasticsearch.server@9.0.0-SNAPSHOT/org.elasticsearch.plugins.PluginsService.loadBundle(PluginsService.java:427)
        at org.elasticsearch.server@9.0.0-SNAPSHOT/org.elasticsearch.plugins.PluginsService.lambda$loadPluginBundles$2(PluginsService.java:219)
        at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:184)
        at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:215)
        at java.base/java.util.Iterator.forEachRemaining(Iterator.java:133)
        at java.base/java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1939)
        at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:570)
        at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:560)
        at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:151)
        at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:174)
        at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:265)
        at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:636)
        at org.elasticsearch.server@9.0.0-SNAPSHOT/org.elasticsearch.plugins.PluginsService.loadPluginBundles(PluginsService.java:219)
        at org.elasticsearch.server@9.0.0-SNAPSHOT/org.elasticsearch.plugins.PluginsService.<init>(PluginsService.java:94)
        at org.elasticsearch.server@9.0.0-SNAPSHOT/org.elasticsearch.node.NodeServiceProvider.newPluginService(NodeServiceProvider.java:56)
        at org.elasticsearch.server@9.0.0-SNAPSHOT/org.elasticsearch.node.NodeConstruction.createEnvironment(NodeConstruction.java:482)
        at org.elasticsearch.server@9.0.0-SNAPSHOT/org.elasticsearch.node.NodeConstruction.prepareConstruction(NodeConstruction.java:276)
        at org.elasticsearch.server@9.0.0-SNAPSHOT/org.elasticsearch.node.Node.<init>(Node.java:185)
        at org.elasticsearch.server@9.0.0-SNAPSHOT/org.elasticsearch.bootstrap.Elasticsearch$2.<init>(Elasticsearch.java:281)
        at org.elasticsearch.server@9.0.0-SNAPSHOT/org.elasticsearch.bootstrap.Elasticsearch.initPhase3(Elasticsearch.java:281)
        at org.elasticsearch.server@9.0.0-SNAPSHOT/org.elasticsearch.bootstrap.Elasticsearch.main(Elasticsearch.java:83)
    Caused by: java.lang.reflect.InvocationTargetException
        at java.base/jdk.internal.reflect.DirectConstructorHandleAccessor.newInstance(DirectConstructorHandleAccessor.java:74)
        at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:501)
        at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:485)
        at org.elasticsearch.server@9.0.0-SNAPSHOT/org.elasticsearch.plugins.PluginsService.loadPlugin(PluginsService.java:514)
        ... 21 more
    Caused by: java.lang.IllegalArgumentException: Failed to parse value [true] for setting [xpack.ml.enabled]
        at org.elasticsearch.server@9.0.0-SNAPSHOT/org.elasticsearch.common.settings.Setting.get(Setting.java:572)
        at org.elasticsearch.server@9.0.0-SNAPSHOT/org.elasticsearch.common.settings.Setting.get(Setting.java:531)
        at org.elasticsearch.ml@9.0.0-SNAPSHOT/org.elasticsearch.xpack.ml.MachineLearning.<init>(MachineLearning.java:781)
        at java.base/jdk.internal.reflect.DirectConstructorHandleAccessor.newInstance(DirectConstructorHandleAccessor.java:62)
        ... 24 more
    Caused by: org.elasticsearch.common.settings.SettingsException: xpack.ml.enabled cannot be set to [true] on [darwin-x86_64]
        at org.elasticsearch.xcore@9.0.0-SNAPSHOT/org.elasticsearch.xpack.core.XPackSettings.lambda$static$0(XPackSettings.java:94)
        at org.elasticsearch.server@9.0.0-SNAPSHOT/org.elasticsearch.common.settings.Setting.get(Setting.java:558)
        ... 27 more
```

I'm trying to limit the run to only run my test (hoping that skipping the ML tests would fix this issue). I'm using the syntax from `TESTING.asciidoc`, so the command I'm trying is:
```bash
./gradlew :x-pack:plugin:yamlRestTest --tests "org.elasticsearch.xpack.test.rest.XPackRestIT" -Dtests.method="test {yaml=counted_keyword/*/*}"
```

But I keep getting
```
No tests found for given includes: [**/*$*.class](exclude rules) [org.elasticsearch.xpack.test.rest.XPackRestIT](--tests filter)
```

I've probably got something wrong in my `-Dtests.method` argument, so I'll keep trying variations on that.

jordan-powers committed:
Add yaml rest test for counted_keyword synthetic source

jordan-powers commented:
Thanks @martijnvg for the help! I swapped JDKs and deleted my `~/.gradle` cache, and I was able to run my yaml test with:
```bash
./gradlew :x-pack:plugin:yamlRestTest --tests "org.elasticsearch.xpack.test.rest.XPackRestIT" -Dtests.method="*counted_keyword*"
```

jordan-powers committed:
Merge remote-tracking branch 'upstream/main' into counted_keyword_synthetic_source

martijnvg commented on x-pack/plugin/mapper-counted-keyword/src/main/java/org/elasticsearch/xpack/countedkeyword/CountedKeywordFieldMapper.java:
```diff
@@ -399,4 +488,16 @@ public FieldMapper.Builder getMergeBuilder() {
+    @Override
+    protected SyntheticSourceSupport syntheticSourceSupport() {
+        var keepMode = sourceKeepMode();
+        if (keepMode.isPresent() && keepMode.get() != SourceKeepMode.NONE) {
```
I think for now we need to remove `keepMode.isPresent() && ` from this if statement. 

The issue currently is that if `index.mapping.synthetic_source_keep` index setting is configured or `synthetic_source_keep` is specified at an object field at a higher level, then we don't know about this. The `sourceKeepMode()` only tells about the `synthetic_source_keep` of the current field.

I think we can fix this in a followup. I think we need to add logic to `MapperBuilderContext` to include what synthetic source keep is based on the index setting or any direct or indirect parent object fields. Let's explore this in a follow up PR, then we can get this PR in, knowing that the optimized synthetic source loader only is used if the `synthetic_source_keep` is set to `none` on a field of type `counted_keyword`.

jordan-powers replied:
I agree that we need more logic to detect if synthetic_source_keep is set somewhere else and inherited here.

But if I just remove the `keepMode.isPresent() &&` from the if-statement, couldn't that potentially raise a `NoSuchElementException` if the keepMode is not specified in the field settings?

Sounds to me like we want the logic for now to only use optimized synthetic source if the `synthetic_source_keep` is explicitly `none`. I'll update the logic to instead be
```suggestion
        if (keepMode.isPresent() == false || keepMode.get() != SourceKeepMode.NONE) {
```

martijnvg replied:
Yes, my code suggestion is incorrect. We should fallback is keepMode isn't present or not set to none, like you suggest here.

jordan-powers commented on x-pack/plugin/mapper-counted-keyword/src/main/java/org/elasticsearch/xpack/countedkeyword/CountedKeywordFieldMapper.java:
```diff
@@ -46,6 +49,7 @@
+import org.elasticsearch.xcontent.XContentBuilder;
 import org.elasticsearch.xcontent.XContentParser;
 
 import java.io.IOException;
```
Oops, missed that, I'll update it!

jordan-powers committed:
Only support synthetic source if keep mode is none

Right now we don't have the logic set up to get the correct value of
synthetic_source_keep if the value is inherited.

Until we get that set up, we can only confiently use the doc_values
implementation of synthetic_source if the synthetic_source_keep is
explicitly set to "none" in the mapping parameters.

Commit:
[CI] Auto commit changes from spotless

jordan-powers committed:
Merge remote-tracking branch 'upstream/main' into counted_keyword_synthetic_source

jordan-powers committed:
Merge remote-tracking branch 'upstream/main' into counted_keyword_synthetic_source

jordan-powers committed:
Remove use of forbidden @Ignore annotation

martijnvg commented on x-pack/plugin/mapper-counted-keyword/src/test/java/org/elasticsearch/xpack/countedkeyword/CountedKeywordFieldMapperTests.java:
```diff
@@ -64,9 +71,103 @@ protected Object generateRandomInputValue(MappedFieldType ft) {
+
+    @Override
+    public void testSyntheticSourceKeepAll() throws IOException {
+        // For now, synthetic source is only supported when "synthetic_source_keep" is "none"
```
```suggestion
        // For now, native synthetic source is only supported when "synthetic_source_keep" mapping attribute is "none"
```

martijnvg reviewed (approved):
LGTM 👍

jordan-powers committed:
Update comment per review

jordan-powers committed:
Merge remote-tracking branch 'upstream/main' into counted_keyword_synthetic_source

kkrik-es commented:
@jordan-powers seems like `auto-backport` didn't run, mind backporting manually?

jordan-powers commented:
Backport PR for 8.x opened, #120541

lkts commented on x-pack/plugin/mapper-counted-keyword/src/test/java/org/elasticsearch/xpack/countedkeyword/CountedKeywordFieldMapperTests.java:
```diff
@@ -64,9 +71,103 @@ protected Object generateRandomInputValue(MappedFieldType ft) {
+            }
+
+            private Tuple<String, String> generateValue() {
+                String v = ESTestCase.randomAlphaOfLength(5);
```
Should we explicitly add some duplicate values here since this is one of the main features of the field? This code would do that with some probability but i think it would be nice to be explicit about it.

jordan-powers replied:
Good point, I'll include the updates in the follow-up PR.
